domain/src/DomainValidator.php

```diff
@@ -26,103 +37,72 @@ class DomainValidator implements DomainValidatorInterface {
     $this->httpClient = $http_client;
+    $this->entityTypeManager = $entity_type_manager;
   }
 
   /**
    * {@inheritdoc}
    *
-   * @TODO: Divide this into separate methods. Do not return Drupal-specific
-   * responses.
+   * @TODO: Do not return Drupal-specific responses.
    */
   public function validate(DomainInterface $domain) {
     $hostname = $domain->getHostname();
-    $error_list = array();
-    // Check for at least one dot or the use of 'localhost'.
-    // Note that localhost can specify a port.
-    $localhost_check = explode(':', $hostname);
-    if (substr_count($hostname, '.') == 0 && $localhost_check[0] != 'localhost') {
-      $error_list[] = $this->t('At least one dot (.) is required, except when using <em>localhost</em>.');
-    }
-    // Check for one colon only.
-    if (substr_count($hostname, ':') > 1) {
-      $error_list[] = $this->t('Only one colon (:) is allowed.');
-    }
-    // If a colon, make sure it is only followed by numbers.
-    elseif (substr_count($hostname, ':') == 1) {
-      $parts = explode(':', $hostname);
-      $port = (int) $parts[1];
-      if (strcmp($port, $parts[1])) {
-        $error_list[] = $this->t('The port protocol must be an integer.');
-      }
-    }
-    // The domain cannot begin or end with a period.
-    if (substr($hostname, 0, 1) == '.') {
-      $error_list[] = $this->t('The domain must not begin with a dot (.)');
-    }
-    // The domain cannot begin or end with a period.
-    if (substr($hostname, -1) == '.') {
-      $error_list[] = $this->t('The domain must not end with a dot (.)');
-    }
+
+    // First, do generalized hostname validation.
+    $error_list = $this->staticHostnameValidation($hostname);
+
     // Check for valid characters, unless using non-ASCII domains.
-    $non_ascii = $this->configFactory->get('domain.settings')->get('allow_non_ascii');
-    if (!$non_ascii) {
-      $pattern = '/^[a-z0-9\.\-:]*$/i';
-      if (!preg_match($pattern, $hostname)) {
-        $error_list[] = $this->t('Only alphanumeric characters, dashes, and a colon are allowed.');
-      }
-    }
-    // Check for lower case.
-    if ($hostname != Unicode::strtolower($hostname)) {
-      $error_list[] = $this->t('Only lower-case characters are allowed.');
+    if (! $this->config->get('allow_non_ascii') && ! preg_match('/^[a-z0-9:.-]*$/i', $hostname)) {
+      $error_list[] = $this->t('Only alphanumeric characters, dashes, and a colon are allowed.');
     }
+
     // Check for 'www' prefix if redirection / handling is
     // enabled under global domain settings.
     // Note that www prefix handling must be set explicitly in the UI.
     // See http://drupal.org/node/1529316 and http://drupal.org/node/1783042
-    if ($this->configFactory->get('domain.settings')->get('www_prefix') && (substr($hostname, 0, strpos($hostname, '.')) == 'www')) {
+    if ($this->config->get('www_prefix') && 0 === strpos($hostname, '.www')) {
```

Comment:
The strpos statement here is backward. The check condition is `www.` not `.www`. I also don't use the "value" first if ordering.